walker.go

```diff
@@ -0,0 +1,429 @@
+package format
+
+// TODO: This can be extracted from the `go-ipld-format` repo.
+
+import (
+	"context"
+	"errors"
+)
+
+// Walker provides methods to move through a DAG of nodes that implement
+// the `NavigableNode` interface. It uses iterative algorithms (instead
+// of recursive ones) that expose the `path` of nodes from the root to
+// the `ActiveNode` it currently points to.
+//
+// It provides a `Visitor` function to perform user specified logic on
```

Comment:
Hmmm, I don’t think wording quite makes sense. It’s the *user* of the Walker that provides the visitor.

Reply:
Yes, the user should provide the `Visitor` implementation, i.e., the logic of what to do with the nodes we're visiting. I meant that the `Walker` provides the `Visitor` definition itself to allow the user that kind of flexibility. But you're right, that is confusing, how would you word it?

Reply:
Maybe:

> It provides multiple ways to walk through the graph (e.g. `Iterate` and `Seek`). When using them, you provide a `Visitor` function that will be called for each node the Walker traverses. The Visitor can read data from those nodes and, optionally, direct the movement of the Walker by calling `Pause()` (to stop traversing and return) or `Next()` (to skip a child and its descendants).
> 
> [I might skip discussion of `up` and `down` here and describe them elsewhere since those are really implementation details.]

Reply:
Three options in my mind:

1. Do you need to describe them collectively at all, or are their own function-level doc comments enough? It feels like the implementations of `Iterate` and `Seek` are very clear and readable, so reviewing how they use `down` and `up` in narrative text instead of code might not be that big a deal.

2. A separate “guide to the implementation” comment at the top of the file, separate from the struct definition.

3. Your suggestion — a loose comment paragraph after the struct definition.

Reply:
I went with number 3, but I also like option 2, I'll keep it in mind for the future.